### DISCUSSION TITLE
`uart.write()` ?randomly? blocks in SAMD51

### DISCUSSION THREAD

**Discussion by cajomferro** [micropython/micropython#10061] [Hardware & Peripherals]
I have an ATSAMD51 microcontroller integrated in a project-specific board which has also a xbee3 RF module. I've built the firmware myself based on the SPARKFUN_SAMD51_THING_PLUS port. The microcontroller is currently running Micropython v1.19.1-712-gb120d3b2f including [this patch](https://github.com/micropython/micropython/pull/10021) which affected the `uart.read()`.

I can access the Microcontroller REPL via UART (over USB port on macos) and I've managed to send/receive commands to the xbee3 via a secondary UART.

However, there seems to be a problem with the `uart.write()` function. Randomly (seems to me) this function blocks the entire microcontroller (cannot use the REPL anymore), and only a button reset can fix it. Moreover, after the reset I have to wait some seconds ( ~10) before I can access the REPL again.

This is a simple demo I've made to try to characterize the problem:

```python
from utime import sleep, sleep_ms
from machine import UART, Pin

uart = UART(5, rx=Pin('XBEE_RX'), tx=Pin('XBEE_TX'), baudrate=115200)

def start(time_ms=500):
    x = 1
    while True:
        print(f"#{x}")
        uart.write('foo')
        sleep_ms(50) # wait a bit for xbee to process info, otherwise nothing will be read
        data = uart.read()
        if data:
            data = data.decode("utf-8").strip()
        print(data)
        x += 1
        sleep_ms(time_ms)
```

I've run this demo using different sleeping durations (50ms, 100ms, and 1000ms). The numbers on the right represent the iteration when the `uart.write()` blocks. So, for example, using `time_ms=50` on the second run, it got 249 iterations before blocking and using `time_ms=100` on the fifth run, it got only 2 iterations before blocking. So it seems to me it is a random behavior right?

| time_ms | iteration where it blocks.    |
---------|-----------------------------|
| 50ms | 69, 249, 143, 65, 207, 360, 4  |
| 100ms | 57, 29, 110, 31, 2, 189, 6         |
| 1000ms | 154, 70, 75, 30, 25, 99, 138 |

Can anyone replicate this situation?

Thank you in advance and congratulations on the Micropython project. It is awesome!!

**Comment by robert-hh** [comment]
Thank you for the report. I have set up the test with a ItsyBitsy M4, but at first I just loop back tx to RX. Not stops within > 1000 iterations for both 50ms and 100ms. I have set up a simple echo test as well (see below) using an rp2 as counterpart with the same results. Only the rp2 seems to pile up incoming messages, if the delay is shorter than 70ms. You may as well try the loopback test to sort out the XBEE.
The 10sec delay after reset is strange. I have never seen this, and I push reset pretty often.

```
from machine import UART, Pin
import time

uart=UART(0, rx=Pin(1), tx=Pin(0))

n=0
while True:
    data = uart.read(3)
    if data:
        print(n, data)
        n += 1
        time.sleep_ms(10)
        uart.write("bar")
```

**Reply by cajomferro** [reply]
I've just tried this example. I've added a `uart.write("bar")` before running the loop, otherwise the first read will not have nothing ever.

For me, using the XBEE it stops after a very few runs (e.g., 43, 75, 65 times). Currently, I don't have any other board with me so I cannot test another board with XBEE. Also, do you have a SPARKFUN_SAMD51_THING_PLUS board? Because I'm using the exact same microcontroller (SAMD51J20A).

So I am guessing there are two (main) options here:
a) there is a specific behavior of the XBEE that somehow affects the `uart.write()` (this could be tested with another board+xbee which I don't have now)
b) there is a problem with the SAMD51J20A port (and this same behavior should happen with XBEE or any other uart device)

Regarding the 10sec delay after reset is probably something host OS related..

Thanks!

**Reply by robert-hh** [reply]
So I have now connected a SPARKFUN_SAMD51_THING_PLUS board. Same result. With the Adafruit board, I stopped at ~17_000. The Sparkfun board is now at ~3000.

The Sparkfun board is sometimes a little bit slow in booting, like about a second startup time. Something I have to check. Maybe the crystal takes too long to get in sync. That may also cause your startup delay. You could check by commenting in boards/SPARKFUN_SAMD51_THING_PLUS/mpconfigboard.h the line:
`#define MICROPY_HW_XOSC32K          (1)'
Without that setting active, the board will use the internal oscillator for the clock.

**Reply by robert-hh** [reply]
The Sparkfun device stopped at iteration 4689! It's interesting that it then does not react to Ctrl-C any more. It should, because in both uart.read() and uart.write() Ctrl-C is checked implicitly. That looks like something hard to find.

**Reply by robert-hh** [reply]
I restarted with the Sparkfun board, and counting is now >20000, still running.

**Reply by cajomferro** [reply]
Ok @robert-hh, I think I have good news (but I'll let you know in case I still encounter any problems). I've tried to comment the MICROPY_HW_XOSC32K line and the results were immediately different. It is now on the 2164 without blocking. Can you explain me better why should I enabling/disabling this line? What about the MICROPY_HW_MCU_OSC32KULP? I saw that line as well on the Feather_M4. In alternative, can you point me to recommended literature on this? Thank you once again! First-class support here!

**Reply by ricksorensen** [reply]
I just ran the simple loop back run on a WIO terminal, using the right hand grove as a serial port (A0/A1- UART(4, rx=Pin("A1_D1"), tx=Pin("A0_D0"), baudrate=115200) and the printout stopped after iteration 5018.  Ctrl-C did not work, needed to reset board.
I changed baudrate=9600 and it is now >181000

MICROPY_HW_XOSC32K is inabled
uart patch is not included
MicroPython v1.15-2129-gfbe02b64b on 2022-11-19; Wio Terminal D51R with SAMD51P19A

as an aside I tried this on a SAMD21 (XIAO) board MICROPY_HW_XOSC32K is inabled her tool
    at 9600 baud it ran >360000 (I stopped it)  UART(4, baudrate=9600, bits=8, parity=None, stop=1, timeout=1, timeout_char=2, rxbuf=256, txbuf=256)
    at baudrate=115200 it stopped after iteration 142.  Ctrl-C did not work, needed to reset board. Ctrl-C did not work, needed to reset board.
MicroPython v1.19.1-678-g8a91c7196 on 2022-11-15; Seeed Xiao with SAMD21G18A

**Reply by robert-hh** [reply]
@cajomferro I had it now running over night with the Sparkfun SAMD51 board. Being at >310_000 cycles. still running.

About `MICROPY_HW_XOSC32K` and `MICROPY_HW_MCU_OSC32KULP`. If both are set, there is a hybrid mode. The peripheral devices' clock is controlled by the crystal, giving precise frequencies for PWM and others, the CPU and memory by the internal oscillator. The Adafruit device has a PCB layout problem, in that the signals for the RGB LED and the Debugger interfere with the crystal itself. When that happened, the board locked up completely. There is some explanation of the settings e.g. in mcu/samd51/clock_config.c.

@ricksorensen Thank you for testing. Please try the PR #10021 patch.

**Reply by robert-hh** [reply]
@cajomferro Neither uart.write() nor uart.read() should block. Both have a timeout check in their send/receive loop.

**Reply by robert-hh** [reply]
@ricksorensen For unknown reason, my SEEED XIAO board refuses to send with UART. It receives, but no data out.I can toggle the Pin with the Pin functions. A Adafruit SAMD21 board works with the same firmware. Whether self-built or taken from the web size. Strange! But looks like another topic. It seems that your board works.

The actual test set-up with the Adafruit SAMD21 board works fine. Now at 14000.

I'm going for my morning walk now. Sometimes it helps to get a clue.
Edit: Problem solved. Located between chair and keyboard.

**Reply by robert-hh** [reply]
Test with the Sparkfun SAMD51 still running, now at 500000. I started the SAMD21 test with a SEEED XIAO board again at 115200 baud, now at 71000. Looks good so far.

**Reply by ricksorensen** [reply]
@robert-hh I have a curious result with both the SAMD51 (WIO Terminal) and SAMD21(XIAO).  Both of them run with no problems when I first start the UART test.  After stopping it with Ctrl-C, and restarting with a new baudrate very quickly (<1000) interations the program stops updating the REPL display and cannot be stopped with Ctrl-C.  

The firmware was built from  `https://github.com/robert-hh/micropython` using `samd_sp2` branch pulled at about 1400 UTC on 20221124

Here is the code for my WIO Terminal runs:
```
from machine import UART, Pin
import time

def dotest(baud=115200, delay=10):
    uart = UART(4,
                rx=Pin("A1_D1"),
                tx=Pin("A0_D0"),
                baudrate=baud)
    n = 0
    uart.write("xxx")   # drive first message
    print(uart)
    print("delay = {}".format(delay))
    try:
        while True:
            data = uart.read(3)
            if data:
                print("{}: {}".format(n, data), end='\r')
                n += 1
                time.sleep_ms(delay)
                uart.write("bar")
    except:
        print("\nstopped after {} iterations".format(n))
```
And her is the terminal output for one example:
```
$mpremote
(Ctrl-D ... output from my boot.py)
MPY: soft reboot
stack: 404 out of 15360
GC: total: 173376, used: 1264, free: 172112
 No. of 1-blocks: 7, 2-blocks: 5, max blk sz: 18, max free sz: 10661
FS:  Total=130560   Free=115200
MicroPython v1.15-2172-g89a0c5c58 on 2022-11-24; Wio Terminal D51R with SAMD51P19A
Type "help()" for more information.
>>> import wiouartchk
>>> wiouartchk.dotest()
UART(4, baudrate=115200, bits=8, parity=None, stop=1, timeout=1, timeout_char=1, rxbuf=256, txbuf=256)
delay = 10
1498: b'bar'
stopped after 1499 iterations
>>> wiouartchk.dotest(baud=9600)
UART(4, baudrate=9600, bits=8, parity=None, stop=1, timeout=1, timeout_char=2, rxbuf=256, txbuf=256)
delay = 10
1152: b'bar'    (no CtrlC, reset WIO Terminal) device disconnected
$ mpremote
Connected to MicroPython at /dev/ttyACM0
Use Ctrl-] to exit this shell

>>> import wiouartchk
>>> wiouartchk.dotest(9600)
UART(4, baudrate=9600, bits=8, parity=None, stop=1, timeout=1, timeout_char=2, rxbuf=256, txbuf=256)
delay = 10
4620: b'bar'
stopped after 4621 iterations
>>> wiouartchk.dotest(115200)
UART(4, baudrate=115200, bits=8, parity=None, stop=1, timeout=1, timeout_char=1, rxbuf=256, txbuf=256)
delay = 10
698: b'bar' (no CtrlC, reset WIO Terminal) device disconnected 
```

After one restart (power cycle) I had to reload the `firmware.uf2` file - but only once.  Probably unrelated (or as you put it chair-computer interface).

**Reply by robert-hh** [reply]
Hmm. That's strange. I have stopped and restarted the XIAO a few times for changing the test code, but that included an interaction with mpremote edit filename. And mpremote goes though a soft reset. Do the WIO and the XIAO talk with each? I use a RP2 for the counterpart.
Could you try to call uart.deinit() before changing the baudrate?
The firmware version should be fine. I did not update the branch today.

**Reply by robert-hh** [reply]
Do both device stop, or only the WIO?

**Reply by robert-hh** [reply]
Ok. I can replicate the behavior with the XIAO/rp2 combo. Lets see if that happens with the ItsyBitsy M4 as well, because that one is connected to the hardware debugger.

**Reply by robert-hh** [reply]
It happens with the ItsyBitsy as well. The code still runs, but rotating through the IRQ handler.

**Reply by robert-hh** [reply]
It could be a dynamic memory problem. You could try to instantiate UART only once and change the baud rate with uart.init().
b.t.w.: You do not have to change the baud rate. Just instantiating UART again with the same parameters does the same.
There was a change some time ago about dynamic memory handling with introducing MP_REGISTER_ROOT_POINTER() and MP_STATE_PORT() calls.

**Reply by ricksorensen** [reply]
My tests were done with an end-around on each machine.. talking to itself.  The same thing occurs when I connected them to each other.  I'll rerun the scripts with a memory allocation check at the beginning, and then using uart.init() to change baud rate.

**Reply by robert-hh** [reply]
Any result so far? I have to scan the allocation of memory and the protection against being freed by gc.collect(). 
The test with the Sparkfun SAMD51 is now at >1.1 Million cycles. I will stop that now. @cajomferro : What is the state of your testing with your own board?

**Reply by robert-hh** [reply]
> My tests were done with an end-around on each machine.. talking to itself.

Therefore it works without the samd_sp2 patches.

**Reply by robert-hh** [reply]
@ricksorensen So I found the reason. The UART object is created with a finalizer. As such, __del__ is called for it when the object went out of scope and garbage collection runs.  Then, uart.deinit() is called, which kind of stopped the code halfways. That is to be fixed. During that analysis, I could as well simplify the code a little bit. The only thing I have to determine is, whether the __del__ code must free the object it's called for.

**Reply by robert-hh** [reply]
@ricksorensen @cajomferro I have update the samd_sp2 branch. It worked well in a few test series of >10000 iterations. The new code is a little bit smaller and simpler.

**Reply by ricksorensen** [reply]
Thanks so much for finding these solutions!  I ran a test program with stressing the reassignment of the UART - 20 reassignments of the same UART config to a function variable on both the SAMD21 XIAO and SAMD51 WIO Terminal - with no failures.   (In both cases it was UART(4))  ... Looking at memory with gc.mem_free() after a gc.collect() there was no change in the available GC memory on either MCU.  

Thanks again.

**Reply by cajomferro** [reply]
@robert-hh thank you for the nice support! I'll let you know my test once I got back to work on it (probably next week).

**Comment by robert-hh** [comment]
Fixed by  #10021 .
